### PR TITLE
HDDS-6863. Add Group-Id & Ratis-Roles Information for OM UI.

### PR DESCRIPTION
**Pull request by ArafatKhan2198**
## What changes were proposed in this pull request?

Add Group-Id, Current-Node Role, Election Count, Leader Election Elapsed time & the Ratis-Ring information for OM UI.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-6863?filter=-1

## How was this patch tested?

Tested on docker OM ha env
image

For a single node cluster it will print STANDALONE 
image

### CODE REVIEW

**Comment by adoroszlai** [comment]
@ArafatKhan2198 I think it is good practice to [ensure CI run in one's own fork succeeds, before opening a PR](https://lists.apache.org/thread/24szm50ybx7ww8zklfzls77p4h9q77yy).

**Comment by ArafatKhan2198** [comment]
> ensure CI run in one's own fork succeeds, before opening a PR

Thanks I was not aware of that !!
I will do that next time :)

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 1]
```diff
@@ -2966,6 +2967,40 @@ public String getRpcPort() {
...
+                info.getOmRoleInfo().getNodeId(),
+                info.getOmRoleInfo().getServerRole()
+            ));
+      }
+    }
+    return sb.toString();
+  }
+
+  @Override
+  public String getCurrentHost() {
```
Comment (sadanand48) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2995]: I don’t think we need to add the hostname, Without knowing the hostname you wouldn’t be able to access UI itself.

File: hadoop-ozone/ozone-manager/src/main/resources/webapps/ozoneManager/ozoneManager.js [hunk 2]
```diff
@@ -113,5 +113,12 @@
         require: {
             overview: "^overview"
         },
+        controller: function ($http) {
+            var ctrl = this;
+            $http.get("http://localhost:9874/jmx?qry=Ratis:service=RaftServer,group=*,id=*")
```
Comment (sadanand48) [inline hadoop-ozone/ozone-manager/src/main/resources/webapps/ozoneManager/ozoneManager.js:118]: Don't need localhost here

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 3]
```diff
@@ -2966,6 +2967,40 @@ public String getRpcPort() {
...
+    StringBuilder sb = new StringBuilder();
+    try {
+      serviceList = getServiceList();
+    } catch (IOException e) {
+      return "";
+    }
+    for (ServiceInfo info : serviceList) {
+      if (info.getOmRoleInfo() != null) {
+        sb.append(
+            String.format(
```
Comment (sadanand48) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2982]: Add the formatting part to a utility method maybe?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 4]
```diff
@@ -2966,6 +2967,40 @@ public String getRpcPort() {
...
 
   @Override
+
+  public String getOmRatisRoles() {
+    List<ServiceInfo> serviceList = null;
+    StringBuilder sb = new StringBuilder();
+    try {
+      serviceList = getServiceList();
+    } catch (IOException e) {
+      return "";
```
Comment (sadanand48) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2977]: Please print a meaningful error LOG message here instead of returning an empty string

File: hadoop-ozone/ozone-manager/src/main/resources/webapps/ozoneManager/om-overview.html [hunk 5]
```diff
@@ -22,6 +22,18 @@ <h2>Status</h2>
         <td>Rpc port</td>
         <td>{{$ctrl.overview.jmx.RpcPort}}</td>
     </tr>
+    <tr>
+        <td>OM Roles (HA)	</td>
+        <td>{{$ctrl.overview.jmx.OmRatisRoles}}</td>
+    </tr>
+    <tr>
+        <td>Current-Host</td>
```
Comment (sadanand48) [inline hadoop-ozone/ozone-manager/src/main/resources/webapps/ozoneManager/om-overview.html:30]: Can  remove current host field

File: hadoop-ozone/ozone-manager/src/main/resources/webapps/ozoneManager/ozoneManager.js (hunk 2 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/ozone-manager/src/main/resources/webapps/ozoneManager/ozoneManager.js:118]: Removed it, thanks !!

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java (hunk 1 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2995]: Sure Ill remove it 
Thanks !

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java (hunk 3 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2982]: Good Idea I'll add the formatting method to `OmUtils.java` class !!

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java (hunk 4 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2977]: Sure will do 
Thanks !!

File: hadoop-ozone/ozone-manager/src/main/resources/webapps/ozoneManager/om-overview.html (hunk 5 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/ozone-manager/src/main/resources/webapps/ozoneManager/om-overview.html:30]: I have removed it and replaced it with current role 
Thanks !!

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 6]
```diff
@@ -2966,6 +2966,19 @@ public String getRpcPort() {
...
 
   @Override
+
+  public String getOmRatisRoles() {
+    List<ServiceInfo> serviceList = null;
+    int port = omNodeDetails.getRatisPort();
+    try {
+      serviceList = getServiceList();
+    } catch (IOException e) {
+      LOG.error("IO-Exception Occurred");
```
Comment (jojochuang) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2976]: better to log the exception
```suggestion
      LOG.error("IO-Exception Occurred", e);
```

Comment (ArafatKhan2198) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2976]: Will do, thanks for the suggestion !!

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 7]
```diff
@@ -2966,6 +2966,19 @@ public String getRpcPort() {
...
+      return "Null values found";
```
[the remark this answers is not in the document]
Comment (ArafatKhan2198) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2977]: Will do, thanks for the suggestion !!

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 8]
```diff
@@ -2966,6 +2966,19 @@ public String getRpcPort() {
   }
 
   @Override
+
+  public String getOmRatisRoles() {
+    List<ServiceInfo> serviceList = null;
+    int port = omNodeDetails.getRatisPort();
+    try {
+      serviceList = getServiceList();
+    } catch (IOException e) {
```
Comment (sadanand48) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2975]: Using getServiceList() won't work because according to it's [logic](https://github.com/apache/ozone/blob/e6b6a842a19e5c64e62e03b039aa05fbe218ddbf/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java#L3023) , the node that calls this method is the leader and since in the web UI case every OM node will make this call, it will show every node as leader

**Review by sadanand48 — COMMENTED** [review]
[Logic](https://github.com/apache/ozone/blob/e6b6a842a19e5c6[Logic ](https://github.com/apache/ozone/blob/e6b6a842a19e5c64e62e03b039aa05fbe218ddbf/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java#L3023) 4e62e[](https://github.com/apache/ozone/blob/e6b6a842a19e5c64e62e03b039aa05fbe218ddbf/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java#L3023) 03b039aa05fbe218ddbf/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java#L3023)

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java (hunk 8 above)
Comment (sadanand48) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2975]: Check this out https://github.com/apache/ozone/pull/3401/files

Comment (ArafatKhan2198) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2975]: Thanks for the help @sadanand48, I have made changes to the patch do take a look!

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java [hunk 9]
```diff
@@ -267,6 +269,11 @@ public static boolean isReadOnly(
     case ListTenant:
     case TenantGetUserInfo:
     case TenantListUser:
+    case RangerBGSync:
+      // RangerBGSync is a read operation in the sense that it doesn't directly
```
Comment (sadanand48) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java:273]: Unrelated change

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java [hunk 10]
```diff
@@ -789,4 +796,28 @@ public static String getOMAddressListPrintString(List<OMNodeDetails> omList) {
     printString.append("]");
     return printString.toString();
   }
+
+  public static String format(List<ServiceInfo> nodes, int port) {
```
Comment (sadanand48) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java:800]: Instead of multiple fields like Current Role, leaderID , we can rather add this info in this output itself, just add a logic like 

```
public static String format(List<ServiceInfo> nodes,
      int port, String leaderId) {
    StringBuilder sb = new StringBuilder();
    int count = 0;
    for (ServiceInfo info : nodes) {
      // Printing only the OM's running
      String role = info.getOmRoleInfo().getNodeId().equals(leaderId)?"LEADER":"FOLLOWER";
      if (info.getNodeType() == HddsProtos.NodeType.OM) {
        sb.append(
            String.format(
                " { HostName: %s | Node-Id: %s | Ratis-Port : %d | Role : %d } ",
                info.getHostname(),
                info.getOmRoleInfo().getNodeId(),
                port,
                role
            ));
```

File: hadoop-ozone/ozone-manager/src/main/resources/webapps/ozoneManager/ozoneManager.js [hunk 11]
```diff
@@ -113,5 +113,12 @@
         require: {
             overview: "^overview"
         },
+        controller: function ($http) {
+            var ctrl = this;
+            $http.get("jmx?qry=Ratis:service=RaftServer,group=*,id=*")
+                .then(function (result) {
```
Comment (sadanand48) [inline hadoop-ozone/ozone-manager/src/main/resources/webapps/ozoneManager/ozoneManager.js:119]: can remove this part if the above approach works

**Review by sadanand48 — COMMENTED** [review]
Thanks @ArafatKhan2198 for addressing the comments. Overall changes look good, dropped some minor comments inline.

File: hadoop-ozone/ozone-manager/src/main/resources/webapps/ozoneManager/ozoneManager.js (hunk 11 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/ozone-manager/src/main/resources/webapps/ozoneManager/ozoneManager.js:119]: We will be needing it to display the group ID and the current role !!

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java (hunk 10 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java]: Thanks, I have added the role-info to the ratis ring !!

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java (hunk 9 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java]: I have removed it 
thanks !!

File: hadoop-ozone/ozone-manager/src/main/resources/webapps/ozoneManager/om-overview.html [hunk 12]
```diff
@@ -22,6 +22,22 @@ <h2>Status</h2>
         <td>Rpc port</td>
         <td>{{$ctrl.overview.jmx.RpcPort}}</td>
     </tr>
+    <tr ng-hide="!$ctrl.overview.jmx.RatisLeader">
+        <td>Leader-Id</td>
```
[the remark this answers is not in the document]
Comment (ArafatKhan2198) [reply hadoop-ozone/ozone-manager/src/main/resources/webapps/ozoneManager/om-overview.html:26]: Sorry missed out on that 
Thanks!

**Review by sadanand48 — APPROVED** [review]
Thanks @ArafatKhan2198 for addressing comments, +1 LGTM

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java [hunk 13]
```diff
@@ -789,4 +791,31 @@ public static String getOMAddressListPrintString(List<OMNodeDetails> omList) {
...
     return printString.toString();
   }
+
+  public static String format(List<ServiceInfo> nodes, int port,
+                              String leaderId) {
+    StringBuilder sb = new StringBuilder();
+    int count = 0;
+    for (ServiceInfo info : nodes) {
+      // Printing only the OM's running
+      String role = info.getHostname().equals(leaderId) ? "LEADER" : "FOLLOWER";
```
Comment (adoroszlai) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java:801]: Shouldn't we compare `info.getOmRoleInfo().getNodeId()` to `leaderId`?  If hostname and node ID are different, currently we get FOLLOWER for all 3 OMs.
Screenshot 2022-07-25 at 14 09 40

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java [hunk 14]
```diff
@@ -789,4 +791,31 @@ public static String getOMAddressListPrintString(List<OMNodeDetails> omList) {
     printString.append("]");
     return printString.toString();
   }
+
+  public static String format(List<ServiceInfo> nodes, int port,
+                              String leaderId) {
+    StringBuilder sb = new StringBuilder();
+    int count = 0;
+    for (ServiceInfo info : nodes) {
```
Comment (adoroszlai) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java:799]: `nodes` may be in any order.  It would be nice if we could ensure ordering is consistent across all OMs.
Screenshot 2022-07-25 at 14 12 59
Screenshot 2022-07-25 at 14 13 06

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMMXBean.java [hunk 15]
```diff
@@ -29,6 +29,8 @@ public interface OMMXBean extends ServiceRuntimeInfo {
 
   String getRpcPort();
 
+  String getOmRatisRoles();
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMMXBean.java:32]: Nit: since `OMMXBean` is specific to OM, could be simplified to `getRatisRoles()` (which is consistent with `getRatisLogDirectory()` below).

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 16]
```diff
@@ -2966,6 +2967,20 @@ public String getRpcPort() {
   }
 
   @Override
+  public String getOmRatisRoles() {
+    List<ServiceInfo> serviceList = null;
+    int port = omNodeDetails.getRatisPort();
+    RaftPeer leaderId;
+    try {
+      leaderId = omRatisServer.getLeader();
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2975]: Should it handle the case when Ratis is not enabled in OM?

**Review by adoroszlai — CHANGES_REQUESTED** [review]
Thanks @ArafatKhan2198 for the patch.  Mostly looks good, but I think there are some edge cases, and some place for improvement.

**Review by jojochuang — COMMENTED** [review]
Looks okay after Attila's comments are addressed.

**Comment by jojochuang** [comment]
For future reference, it was be nice to show current term and how long has pasted since the last election. (Useful to know if there was any recent election and whether or not the OMs were unstable)

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMMXBean.java (hunk 15 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMMXBean.java]: Changed, thanks for the suggestion !!

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java (hunk 16 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java]: I have added a check for Ratis!! 
thanks for the suggestion !!

File: hadoop-ozone/ozone-manager/src/main/resources/webapps/ozoneManager/ozoneManager.js [hunk 17]
```diff
@@ -113,5 +113,36 @@
...
+
+            $http.get("jmx?qry=ratis:name=ratis.leader_election.*lastLeaderElectionElapsedTime")
+                .then(function (result) {
+                    ctrl.elapsedTime = result.data.beans[0];
+                    ctrl.elapsedTime.Value = convertMsToTime(ctrl.elapsedTime.Value);
+                });
+        }
     });
+        function convertMsToTime(ms) {
+          if(ms == -1){
```
Comment (sadanand48) [inline hadoop-ozone/ozone-manager/src/main/resources/webapps/ozoneManager/ozoneManager.js:136]: Is this a valid case? ms==-1 or should we remove it?

**Comment by adoroszlai** [comment]
Thanks @ArafatKhan2198 for updating the patch.  The order of nodes is still "random", the rest looks good.

File: hadoop-ozone/ozone-manager/src/main/resources/webapps/ozoneManager/ozoneManager.js (hunk 17 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/ozone-manager/src/main/resources/webapps/ozoneManager/ozoneManager.js]: Yes I have addressed the issue!
thanks !!

**Comment by ArafatKhan2198** [comment]
@adoroszlai the OM's are now printed in order

**Review by adoroszlai — APPROVED** [review]
Thanks @ArafatKhan2198 for updating the patch with sorting.

**Comment by adoroszlai** [comment]
Note: Although we used to recommend pushing a new commit to re-trigger the tests, Github now allows committers to re-run only the failed jobs. This also prevents the case where previously passed checks fail in a new run.

**Comment by adoroszlai** [comment]
Thanks @ArafatKhan2198 for the patch, @jojochuang, @sadanand48 for the review.
